designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,808 @@
+        },
+    }
+    rules: {
+        "example/rule1": "error"
+    }
+};
+```
+
+This example imports the `ruleNamespaces` from a plugin directly into the same section in the user config.
+
+#### Referencing Parsers and Processors
+
+In `.eslintrc`, the `parser` and `processor` keys required strings to be specified, such as:
+
+```yaml
+plugins: ["markdown"]
+parser: "babel-eslint"
+processor: "markdown/markdown"
+```
+
+In `eslint.config.js`, you would need to pass the references directly, such as:
+
+```js
+exports.config = {
+    files: ["*.js"],
+    parser: require("babel-eslint"),
+    processor: require("eslint-plugin-markdown").processors.markdown
+};
+```
+
+In both cases, users now must pass a direct object reference. This has the benefit of using the builtin Node.js module resolution system or allowing users to specify their own. There is never a question of where the modules will be resolved from.
+
+#### Applying an Environment
+
+Unlike with `.eslintrc` files, there is no `env` key in `eslint.config.js`. Users can mimic the behavior of `env` by assigning directly to the `globals` key:
+
+```js
+const globals = require("globals");
+
+exports.config = {
+    files: ["*.js"],
+    globals: {
+        MyGlobal: true,
+        ...globals.browser
+    }
+};
+```
+
+This effectively duplicates the use of `env: { browser: true }` in ESLint.
+
+**Note:** This would allow us to stop shipping environments in ESLint. We could just tell people to use `globals` in their config and allow them to specify which version of `globals` they want to use.
+
+#### Overriding Configuration Based on File Patterns
+
+Whereas `.eslintrc` had an `overrides` key that made a hierarchical structure, the `eslint.config.js` file does not have any such hierarchy. Instead, users can return an array of configs that should be used. For example, consider this `.eslintrc` config:
+
+```yaml
+plugins: ["react"]
+rules:
+    react/jsx-uses-react: error
+    semi: error
+
+overrides:
+    - files: "*.md"
+      plugins: ["markdown"],
+      processor: "markdown/markdown"
+```
+
+This can be written in `eslint.config.js` as an array of two configs:
+
+```js
+exports.config = [
+    {
+        files: "*.js",
+        defs: {
+            ruleNamespaces: {
+                react: require("eslint-plugin-react").rules,
+            },
+        }
+        rules: {
+            "react/jsx-uses-react": "error",
+            semi: "error"
+        }
+    },
+    {
+        files: "*.md",
+        processor: require("eslint-plugin-markdown").processors.markdown
+    }
+];
+```
+
+When ESLint uses this config, it will check each `files` pattern to determine which configs apply. Any config with a `files` pattern matching the file to lint will be extracted and used (if multiple configs match, then those configs are merged to determine the final config to use). In this way, returning an array acts exactly the same as the array in `overrides`.
```

Comment:
How do this reconcile differences between files patterns in shared configs?  For example, let's say I pull in a shared config that specifies `files: "*.js"`, as a developer, would I understand why those rules weren't also applying to my `.jsx` files as well?  Should there be guidance to config developers about whether or not they should have `files` in their configs?  With the example above, with both markdown and js sections, which section do rules without explicit `files` get applied?

Reply:
In general, I think most shared configs will either a) not specify `files` or b) if they do specify `files`, indicate that so in their README. 

So yes, I think we'd want to provide guidance to config developers to make sure they specify which files their configs will apply to by default.